File: jafu/src/main/java/org/springframework/fu/jafu/r2dbc/H2R2dbcDsl.java
package org.springframework.fu.jafu.r2dbc;

import java.util.function.Consumer;

import org.springframework.boot.autoconfigure.data.r2dbc.H2DatabaseClientInitializer;
import org.springframework.boot.autoconfigure.data.r2dbc.H2R2dbcProperties;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;
import org.springframework.fu.jafu.AbstractDsl;

/**
 * Jafu DSL for R2DBC H2 configuration.
 *
 * Enable and configure R2DBC support by registering a {@link org.springframework.data.r2dbc.function.DatabaseClient}.
 *
 * Required dependencies are {@code io.r2dbc:r2dbc-h2} and {@code org.springframework.data:spring-data-r2dbc}.
 *
 * @author Sebastien Deleuze
 */
public class H2R2dbcDsl extends AbstractDsl {

	private final Consumer<H2R2dbcDsl> dsl;

	private final H2R2dbcProperties properties = new H2R2dbcProperties();

	H2R2dbcDsl(Consumer<H2R2dbcDsl> dsl) {
		this.dsl = dsl;
	}

	/**
	 * Configure R2DBC H2 support with default properties.
	 * @see org.springframework.fu.jafu.ConfigurationDsl#enable(ApplicationContextInitializer)
	 * @see org.springframework.data.r2dbc.function.DatabaseClient
	 */
	public static ApplicationContextInitializer<GenericApplicationContext> r2dbcH2() {
		return new H2R2dbcDsl(mongoDsl -> {});
	}

	/**
	 * Configure R2DBC H2 support with customized properties.
	 * @see org.springframework.fu.jafu.ConfigurationDsl#enable(ApplicationContextInitializer)
	 * @see org.springframework.data.r2dbc.function.DatabaseClient
	 */
	public static ApplicationContextInitializer<GenericApplicationContext> r2dbcH2(Consumer<H2R2dbcDsl> dsl) {
		return new H2R2dbcDsl(dsl);
	}

	/**
	 * Configure the database url, {@code mem:test;DB_CLOSE_DELAY=-1} by default.
	 * Includes everything after the {@code jdbc:h2:} prefix.
	 * For in-memory and file-based databases, must include the proper prefix (e.g. {@code file:} or {@code mem:}).
	 *
	 * See <a href="http://www.h2database.com/html/features.html#database_url">http://www.h2database.com/html/features.html#database_url</a> for more details.
	 */
	public H2R2dbcDsl url(String url) {
		properties.setUrl(url);
		return this;
	}

	/**
	 * Configure the username.
	 */
	public H2R2dbcDsl username(String username) {
		properties.setUsername(username);
		return this;
	}

	/**
	 * Configure the password.
	 */
	public H2R2dbcDsl password(String password) {
		properties.setPassword(password);
		return this;
	}

	@Override
	public void initialize(GenericApplicationContext context) {
		super.initialize(context);
		this.dsl.accept(this);
		if (properties.getUrl() == null) {
			properties.setUrl("mem:test;DB_CLOSE_DELAY=-1");
		}
		new H2DatabaseClientInitializer(properties).initialize(context);
	}
}

File: kofu/src/main/kotlin/org/springframework/fu/kofu/r2dbc/H2R2dbcDsl.kt
package org.springframework.fu.kofu.r2dbc

import org.springframework.boot.autoconfigure.data.r2dbc.H2DatabaseClientInitializer
import org.springframework.boot.autoconfigure.data.r2dbc.H2R2dbcProperties
import org.springframework.context.support.GenericApplicationContext
import org.springframework.fu.kofu.AbstractDsl
import org.springframework.fu.kofu.ConfigurationDsl

/**
 * Kofu DSL for R2DBC H2 configuration.
 *
 * Enable and [configure][H2R2dbcDsl] R2DBC support by registering a [org.springframework.data.r2dbc.core.DatabaseClient]
 * bean.
 *
 * Required dependencies are `io.r2dbc:r2dbc-h2` and `org.springframework.data:spring-data-r2dbc`.

 * @sample org.springframework.fu.kofu.samples.r2dbcH2
 * @author Sebastien Deleuze
 */
class H2R2dbcDsl(private val init: H2R2dbcDsl.() -> Unit) : AbstractDsl() {

	/**
	 * Configure the database url, `mem:test;DB_CLOSE_DELAY=-1` by default.
	 * Includes everything after the `jdbc:h2:` prefix.
	 * For in-memory and file-based databases, must include the proper prefix (e.g. `file:` or `mem:`).
	 *
	 * See [http://www.h2database.com/html/features.html#database_url](http://www.h2database.com/html/features.html#database_url) for more details.
	 */
	var url: String = "mem:test;DB_CLOSE_DELAY=-1"

	/**
	 * Configure the username.
	 */
	var username: String? = null

	/**
	 * Configure the password.
	 */
	var password: String? = null

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		init()

		val properties = H2R2dbcProperties()
		properties.url = url
		properties.username = username
		properties.password = password

		H2DatabaseClientInitializer(properties).initialize(context)
	}
}

/**
 * Configure R2DBC H2 support.
 * @see H2R2dbcDsl
 */
fun ConfigurationDsl.r2dbcH2(dsl: H2R2dbcDsl.() -> Unit = {}) {
	H2R2dbcDsl(dsl).initialize(context)
}